lib/src/workspace.rs

```diff
@@ -152,45 +159,58 @@ impl Workspace {
         git_repo_path: &Path,
     ) -> Result<(Self, Arc<ReadonlyRepo>), WorkspaceInitError> {
         Self::init_with_backend(user_settings, workspace_root, |store_path| {
-            Box::new(GitBackend::init_external(store_path, git_repo_path))
+            Ok(Box::new(GitBackend::init_external(
+                store_path,
+                git_repo_path,
+            )?))
         })
     }
 
     pub fn init_with_factories(
         user_settings: &UserSettings,
         workspace_root: &Path,
-        backend_factory: impl FnOnce(&Path) -> Box<dyn Backend>,
+        backend_factory: impl FnOnce(&Path) -> Result<Box<dyn Backend>, BackendError>,
         op_store_factory: impl FnOnce(&Path) -> Box<dyn OpStore>,
         op_heads_store_factory: impl FnOnce(&Path) -> Box<dyn OpHeadsStore>,
         index_store_factory: impl FnOnce(&Path) -> Box<dyn IndexStore>,
     ) -> Result<(Self, Arc<ReadonlyRepo>), WorkspaceInitError> {
         let jj_dir = create_jj_dir(workspace_root)?;
-        let repo_dir = jj_dir.join("repo");
-        std::fs::create_dir(&repo_dir).context(&repo_dir)?;
-        let repo = ReadonlyRepo::init(
-            user_settings,
-            &repo_dir,
-            backend_factory,
-            op_store_factory,
-            op_heads_store_factory,
-            index_store_factory,
-        )?;
-        let (working_copy, repo) = init_working_copy(
-            user_settings,
-            &repo,
-            workspace_root,
-            &jj_dir,
-            WorkspaceId::default(),
-        )?;
-        let repo_loader = repo.loader();
-        let workspace = Workspace::new(workspace_root, working_copy, repo_loader)?;
-        Ok((workspace, repo))
+        (|| {
+            let repo_dir = jj_dir.join("repo");
+            std::fs::create_dir(&repo_dir).context(&repo_dir)?;
+            let repo = ReadonlyRepo::init(
+                user_settings,
+                &repo_dir,
+                backend_factory,
+                op_store_factory,
+                op_heads_store_factory,
+                index_store_factory,
+            )
+            .map_err(|repo_init_err| match repo_init_err {
+                RepoInitError::Backend(err) => WorkspaceInitError::Backend(err),
+                RepoInitError::Path(err) => WorkspaceInitError::Path(err),
+            })?;
+            let (working_copy, repo) = init_working_copy(
+                user_settings,
+                &repo,
+                workspace_root,
+                &jj_dir,
+                WorkspaceId::default(),
+            )?;
+            let repo_loader = repo.loader();
+            let workspace = Workspace::new(workspace_root, working_copy, repo_loader)?;
+            Ok((workspace, repo))
+        })()
+        .map_err(|err| {
+            let _ = delete_jj_dir(workspace_root);
+            err
```

Comment:
nit: `delete_jj_dir()` could be inlined as we already know the `jj_dir`. I'm fine with either way.

Reply:
Done.

Reply:
I meant `std::fs::remove_dir_all(jj_dir)`.

Reply:
Done.